plugins/node/instrumentation-amqplib/src/amqplib.ts

```diff
@@ -0,0 +1,733 @@
+    original: Function
+  ) {
+    const self = this;
+    return function ack(
+      this: InstrumentationConsumeChannel,
+      message: amqp.Message,
+      allUpToOrRequeue?: boolean,
+      requeue?: boolean
+    ): void {
+      const channel = this;
+      // we use this patch in reject function as well, but it has different signature
+      const requeueResolved =
+        endOperation === EndOperation.Reject ? allUpToOrRequeue : requeue;
+
+      const spansNotEnded: { msg: amqp.Message }[] =
+        channel[CHANNEL_SPANS_NOT_ENDED] ?? [];
+      const msgIndex = spansNotEnded.findIndex(
+        msgDetails => msgDetails.msg === message
+      );
+      if (msgIndex < 0) {
+        // should not happen in happy flow
+        // but possible if user is calling the api function ack twice with same message
+        self.endConsumerSpan(
+          message,
+          isRejected,
+          endOperation,
+          requeueResolved
+        );
+      } else if (endOperation !== EndOperation.Reject && allUpToOrRequeue) {
+        for (let i = 0; i <= msgIndex; i++) {
+          self.endConsumerSpan(
+            spansNotEnded[i].msg,
+            isRejected,
+            endOperation,
+            requeueResolved
+          );
+        }
+        spansNotEnded.splice(0, msgIndex + 1);
+      } else {
+        self.endConsumerSpan(
+          message,
+          isRejected,
+          endOperation,
+          requeueResolved
+        );
+        spansNotEnded.splice(msgIndex, 1);
+      }
+      return original.apply(this, arguments);
+    };
+  }
+
+  protected getConsumePatch(
+    moduleVersion: string | undefined,
+    original: Function
+  ) {
+    const self = this;
+    return function consume(
+      this: InstrumentationConsumeChannel,
+      queue: string,
+      onMessage: (msg: amqp.ConsumeMessage | null) => void,
+      options?: amqp.Options.Consume
+    ): Promise<amqp.Replies.Consume> {
+      const channel = this;
+      if (
+        !Object.prototype.hasOwnProperty.call(channel, CHANNEL_SPANS_NOT_ENDED)
+      ) {
+        if (self._config.consumeTimeoutMs) {
+          const timer = setInterval(() => {
+            self.checkConsumeTimeoutOnChannel(channel);
+          }, self._config.consumeTimeoutMs);
+          timer.unref();
+          channel[CHANNEL_CONSUME_TIMEOUT_TIMER] = timer;
+        }
+        channel[CHANNEL_SPANS_NOT_ENDED] = [];
+      }
+
+      const patchedOnMessage = function (
+        this: unknown,
+        msg: InstrumentationMessage | null
+      ) {
+        // msg is expected to be null for signaling consumer cancel notification
+        // https://www.rabbitmq.com/consumer-cancel.html
+        // in this case, we do not start a span, as this is not a real message.
+        if (!msg) {
+          return onMessage.call(this, msg);
+        }
+
+        const headers = msg.properties.headers ?? {};
+        const parentContext = propagation.extract(context.active(), headers);
```

Comment:
Are you sure that `context.active()` is used here? in case the headers include a span it's used, otherwise we might link to whatever is active at this time.
Shouldn't we use `ROOT_CONTEXT` here instead to create a new root in case nothing is in the headers?

Reply:
Good catch, I agree about that. Changed to use `ROOT_CONTEXT`